content/en/blog/_posts/2021-06-14-a-better-way-to-provision-kubernetes-resources-using-terraform.md

```diff
@@ -0,0 +1,339 @@
+---
+layout: blog
+title: 'A Better Way to Provision Kubernetes Resources Using Terraform'
+date: 2021-06-14
+slug: a-better-way-to-provision-kubernetes-resources-using-terraform
+---
+
+**Author:** Philipp Strube (Kubestack)
+
+After months of work and thousands of lines of Go, HCL and YAML I am very excited to share the release of the new [Kustomize based Terraform modules for Kubernetes](https://www.kubestack.com/catalog).
+The modules use the [Kustomization provider](https://registry.terraform.io/providers/kbst/kustomization/latest) I maintain as part of my work on the [Kubestack framework](https://www.kubestack.com/).
+Kubestack is an open-source Terraform framework and strives to deliver the best developer experience for teams working with Terraform and Kubernetes.
+
+I believe the new modules are a better way to provision Kubernetes resources using Terraform, because they bring three distinct benefits over the existing alternatives:
+
+1. Like Kustomize, the upstream YAML is left untouched, meaning upstream updates are low maintenance effort
+2. By defining the Kustomize overlay in HCL, all Kubernetes resources are fully customizeable using values from Terraform
+3. Each Kubernetes resources is tracked individually in Terraform state, so diffs and plans show the changes to the actual Kubernetes resources
+
+To make these benefits less abstract, lets compare the [Nginx ingress module](https://www.kubestack.com/catalog/nginx) with [one using the Helm provider](https://registry.terraform.io/modules/terraform-module/release/helm/latest) to provision [Nginx ingress](https://github.com/kubernetes/ingress-nginx).
+
+The Terraform configuration for both examples is available in [this repository](https://github.com/kbst/terraform-helm-vs-kustomize). Lets take a look at the Helm module first.
+
+## The Helm based module
+
+Usage of the module is straightforward. First, we configure the Kubernetes and Helm providers.
+
+```hcl
+provider "kubernetes" {
+  config_path    = "~/.kube/config"
+}
+
+provider "helm" {
+  kubernetes {
+    config_path = "~/.kube/config"
+  }
+}
+```
+
+Then we define a `kubernetes_namespace` and call the `release/helm` module.
+
+```hcl
+resource "kubernetes_namespace" "nginx_ingress" {
+  metadata {
+    name = "ingress-nginx"
+  }
+}
+
+module "nginx_ingress" {
+  source  = "terraform-module/release/helm"
+  version = "2.6.10"
+
+  namespace  = kubernetes_namespace.nginx_ingress.metadata.0.name
+  repository =  "https://kubernetes.github.io/ingress-nginx"
+
+  app = {
+    name          = "ingress-nginx"
+    version       = "3.33.0"
+    chart         = "ingress-nginx"
+    force_update  = true
+    wait          = false
+    recreate_pods = false
+    deploy        = 1
+  }
+
+  set = [
+    {
+      name = "replicaCount"
+      value = 2
+    }
+  ]
+}
+```
+
+If we now run a `terraform plan` for this configuration, we will see the resources to be created.

```

Comment:
To subtly increase the power of your writing, look for "will" and try to replace it with the present-tense.
> If we now run a `terraform plan` for this configuration, we will see the resources to be created.
becomes
> If we now run a `terraform plan` for this configuration, we see the resources to be created.